### ISSUE TITLE
Vert.X Gateway - Transfert-Encoding: chunked with null body on Get Request

### ISSUE THREAD

**Issue by OSMOSA44** [apiman/apiman#859]
Bonjour,

Some backend service, especially Microsoft IIS/10.0 doesn't like to receive a GET Request with a **null body** and a Header "**Transfert-Encodingg: chunked**", the sending response is a 502 Bad Gateway in this case,

To get around this problem, for testing, i added the header "Content-Length: 0" to the request, and the sending response is 200 OK, but this header is not required if there is no body ( cf: [https://tools.ietf.org/html/rfc7230#section-3.3.2](https://tools.ietf.org/html/rfc7230#section-3.3.2) )

If here is a Bug (probably !?), we should adapt the snippet code below and test if a body content is present before set the chunk

[https://github.com/apiman/apiman/blob/5b87973ed2e60ef8ff824c142893f8a4cb3fa179/gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/connector/HttpConnector.java](https://github.com/apiman/apiman/blob/5b87973ed2e60ef8ff824c142893f8a4cb3fa179/gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/connector/HttpConnector.java)
```java
        if (options.hasDataPolicy() || !apiRequest.getHeaders().containsKey("Content-Length")) {
            clientRequest.headers().remove("Content-Length");
            clientRequest.setChunked(true);
        }
```

**Nota:** The problem doesn't appear on Wildfly Gateway,

Regards,

Laurent

**Comment by msavy** [comment]
It was quite a while ago I wrote this stuff, but...

If I recall correctly, this one can be a bit of a tricky situation 🤔, because we don't always know whether there will be a body sent in this situation until we've already sent the headers to the backend API (i.e. the data policy usually doesn't even start transmitting until the connection to the backend has been established and the headers have been sent).

For example, you could have a data policy that decides not to send anything, but by then you've already sent your headers.

Wonder if there's a workaround that we could use.

**Comment by msavy** [comment]
Maybe we should do something special if it's a GET request? There's nothing stopping you having a body with GET, even though it's unusual.

**Comment by msavy** [comment]
Out of interest, what happens if you send a dummy payload or something? For example, just a single space character (in the case that no chunks are ever transmitted and it's a GET).

**Comment by msavy** [comment]
@OSMOSA44 I've made a PoC for you to test, please try it out:

https://github.com/msavy/apiman/tree/iis-bug-idea

In particular:
https://github.com/msavy/apiman/commit/2bfc684f150467103d2d5a5f8b21eef07565d516

I have nothing to test it against, so it's just an idea.

**Comment by msavy** [comment]
I have a couple of other ideas also I'd like you to try out, so please let me know how you get on. Another idea I had was to send a trailer with `Content-Length: 0`.

**Comment by OSMOSA44** [comment]
Hi @msavy , sorry for my late response...

If we know at the end of writing request (after pass all policies)

- that it is a GET request method
- and that the body is empty

We could unset the chunked encoding transfert mode in the `end()` method

```java
    @Override
    public void end() {
        // Try workaround for IIS bug: when chunked encoding is enabled on `GET`, but body ends up being empty
        // we send a character to try to keep it happy.
        if (!anyChunksWritten && this.httpMethod.equals(HttpMethod.GET)) {
            clientRequest.setChunked(false);
        }
        clientRequest.end();
        inboundFinished = true;
    }
```

For test purpose, i have created a test api you can access it
Client -> ILB -> VertX.GW -> **Azure AppGW (IIS)** -> Wildfly.GW -> echo api
https://portail5.api-np.sncf.fr/test/echo/p5p4

For your idea about dummy payload, the request pass correctly through IIS but there is a side-effect if VertX.GW call a Wildfly.GW

```json
{
    "responseCode": 500,
    "message": "io.apiman.gateway.engine.beans.exceptions.ConnectorException: method does not support a request body: GET",
}
```

And add a **Content-Length: 0 without body** on a GET request method seems to be a bad approach for me

Regards,

Laurent

**Comment by msavy** [comment]
I don't think you can unset the chunked mode, because the headers (which includes the chunked mode header) have already been set and sent by the time "end" is called.

The process is:

(1). Connect to back end with headers, including whether it's chunked or content-size
(2). Send body chunks, if any
(3). Call end and finish and/or send trailers.

So by the time you reach (3) it's too late.

**Comment by msavy** [comment]
Re: the dummy payload; in theory GET does actually support a body. So that might be a bug!
